### PR TITLE
mgr/dashboard: smart data for devices with scsi protocol

### PR DESCRIPTION
**Pull request by nizamial09**
In the dashboard, we've been showing smart data for hdd devices with ata
protocol only. Otherwise we show a No Smart Data found error which is
clearly misleading since Smart Data is returned even in the api call.

So this PR is trying to show the smart data for hdd devices
that uses scsi protocol too.

**BEFORE**
before_smart-data

**AFTER**
after2
Screenshot from 2022-05-10 14-38-16

Fixes: https://tracker.ceph.com/issues/55574
Signed-off-by: Nizamudeen A <nia@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/shared/smart-list/fixtures/smart_data_version_1_0_hdd_scsi_response.json [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+          "  number of phys = 1",
+          "  phy identifier = 0",
+          "    attached device type: expander device",
+          "    attached reason: loss of dword synchronization",
+          "    reason: unknown",
+          "    negotiated logical link rate: phy enabled; 12 Gbps",
+          "    attached initiator port: ssp=0 stp=0 smp=1",
+          "    attached target port: ssp=0 stp=0 smp=1",
+          "    SAS address = 0x5000cca2c3d5f545",
+          "    attached SAS address = 0x5000ccab041140bf",
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/shared/smart-list/fixtures/smart_data_version_1_0_hdd_scsi_response.json:155]: Could these be unique IDs and/or sensitive info?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/shared/smart-list/fixtures/smart_data_version_1_0_hdd_scsi_response.json [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+          "errors_corrected_by_eccfast": 0,
+          "errors_corrected_by_rereads_rewrites": 0,
+          "gigabytes_processed": "146241.629",
+          "total_errors_corrected": 0,
+          "total_uncorrected_errors": 0
+        }
+      },
+      "scsi_grown_defect_list": 0,
+      "scsi_version": "SPC-5",
+      "serial_number": "3RKTNS9A",
```
[the remark this answers is not in the document]
Comment (ljflores) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/shared/smart-list/fixtures/smart_data_version_1_0_hdd_scsi_response.json:69]: @epuertat @nizamial09 yes, Telemetry considers host names, device ids, and device serial numbers to be sensitive information. Telemetry's approach is to use the Python `uuid` library to anonymize the host name and the device id. As for the device serial number, Telemetry marks it as `"serial_number": "deleted",` if it detects one.

cc @yaarith

File: src/pybind/mgr/dashboard/frontend/src/app/shared/models/smart.ts [hunk 3]
```diff
@@ -148,6 +158,14 @@ export interface HddSmartDataV1 extends SmartCtlBaseDataV1 {
...
   smartctl: SmartCtlOutput;
   wwn: { id: number; naa: number; oui: number };
+  scsi_error_counter_log: {
+    read: RVWAttributes[];
+  };
+  scsi_grown_defect_list: number;
+  power_on_time: {
+    hours: number;
+    minutes: number;
+  };
```
[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/models/smart.ts]: iSCSI-only. I created a separate interface for them only.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/shared/smart-list/fixtures/smart_data_version_1_0_hdd_scsi_response.json (hunk 1 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/shared/smart-list/fixtures/smart_data_version_1_0_hdd_scsi_response.json]: Changed them! thanks

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/shared/smart-list/smart-list.component.ts [hunk 4]
```diff
@@ -49,7 +50,7 @@ export class SmartListComponent implements OnInit, OnChanges {
   }
 
   isHddSmartData(data: any): data is HddSmartDataV1 {
-    return _.get(data, 'device.protocol', '').toLowerCase() === 'ata';
+    return ['scsi', 'ata'].includes(_.get(data, 'device.protocol', '').toLowerCase());
```
[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/shared/smart-list/smart-list.component.ts]: Yup, I created separate models and functions to take care of that.

**Review by aaSharma14 — APPROVED** [review]

**Review by avanthakkar — APPROVED** [review]

**Comment by avanthakkar** [comment]
jenkins test dashboard

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/shared/smart-list/fixtures/smart_data_version_1_0_hdd_scsi_response.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (yaarith) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/shared/smart-list/fixtures/smart_data_version_1_0_hdd_scsi_response.json]: > If this a serial number from a real device, we shouldn't publish it, or at least we should replace it here with a dummy number. If I'm not wrong it's considered sensitive information

@epuertat I guess you mean this very instance here, in which case is good practice to anonymize sample output.
The real output from the command should allow the user to see in their Ceph Dashboard the actual serial numbers of the devices in their cluster, of course.

[the remark this answers is not in the document]
Comment (epuertat) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/shared/smart-list/fixtures/smart_data_version_1_0_hdd_scsi_response.json]: Thanks @yaarith and @ljflores !

**Pull request merged**